### ISSUE TITLE
Forward-backward test

### ISSUE THREAD

**Issue by utsav-akhaury** [utsav-akhaury/ADMM#5]
To investigate the problems arising from the ADMM + U-Net Plug&Play (see issue #4), I implemented a simple [forward-backward splitting algorithm](https://github.com/utsav-akhaury/ADMM/blob/main/FB_UNet.ipynb) with a single gradient descent step instead of ADMM.

The update equations and derivations are as follows - 

![Screenshot 2021-03-29 at 14 48 47 (2)](https://user-images.githubusercontent.com/59925371/112839049-07801880-909e-11eb-9eba-9bbc3e0964af.png)

For calculating the standard deviation (STD) for all cases, a patch of size 20 X 20 was extracted from the images such that they exclude the region containing the central blob (and only contain the noise map). The following results were obtained after 1 gradient descent step (without denoising) -
- STD [ σ ]  =  23.59   (σ = added noise)
- STD [ n ]  =  22.703041   (noise estimated by n = H x<sub>true</sub> - y)
- STD [ x<sup>(1/2)</sup> ]  =  4.0511665   (observed value)
- STD [ α H<sup>T</sup> n ]  =  4.0120854   (empirical value)
- [ α || H ||<sub>F</sub> σ ]  =  3.8409884   (theoretical value)

![image](https://user-images.githubusercontent.com/59925371/112842465-dd305a00-90a1-11eb-8600-490586272bf3.png)

Subsequently, the input & output to the U-Net must be scaled according to the theoretical value to achieve proper results.

**Comment by utsav-akhaury** [comment]
Instead of the U-Net, I tried using the [Starlet Transform](http://jstarck.free.fr/Chapter_Starlet2011.pdf) as a proximal operator for denoising ([code](https://github.com/utsav-akhaury/ADMM/blob/main/starlet.py)). The results are shown in this [notebook](https://github.com/utsav-akhaury/ADMM/blob/main/FB_starlet.ipynb).

By deconvolving by the telescope PSF, we obtain an image with an infinite resolution. However we don't have ground truth images with infinite resolution. The only ground truths we have are images with HST resolution. In practice, we need to divide the CFHT PSF by the HST PSF in the Fourier space to go from CFHT to HST. As a result, the deconvolution with the partial PSF (which is smaller than the CFHT one) makes the problem less ill-posed and gives better results.

The [Starlet Transform](https://github.com/utsav-akhaury/ADMM/blob/main/starlet.py) code was modified to include the PSF contribution (by convolving the starlet scales with the H<sup>T</sup> operator - the PSF rotated by 180<sup>o</sup>).

When coupled with the forward-backward algorithm, the results show a negative dip around the central blob, which keeps on increasing in magnitude with the no. of iterations.
![image](https://user-images.githubusercontent.com/59925371/113423570-0537fa00-93cf-11eb-9e60-907a32855e02.png)

The negative blob was eliminated by projecting all negative values in the final reconstruction to zero. The galaxy seems to be better reconstructed, although some artefacts/residuals still appear. However, with a simple forward-backward algorithm involving a single gradient-descent step, the reconstruction is not expected to be very accurate. Our aim was just to check whether it performs well when coupled with starlet denoising instead of a U-Net.
![image (1)](https://user-images.githubusercontent.com/59925371/113423655-2862a980-93cf-11eb-9cda-45586ff54e19.png)

**Comment by utsav-akhaury** [comment]
After making the following modifications, the [Forward Backward + Starlet code](https://github.com/utsav-akhaury/ADMM/blob/main/FB_starlet.ipynb) seems to converge -

1. Removing all complex intermediate variables and only dealing with real numbers
2. Computing the Spectral Norm of the gradient Hᵀ.H  (instead of H)
3. Dividing the gradient step size computed using the Spectral Norm by a factor of 5
4. Using Soft thresholding instead of Hard thresholding in Starlet transform
5. Adding the contribution of the Starlet term to the cost function

Without dividing the gradient step size by any factor, the cost function plot is less smooth (which is expected), although there is convergence.
![image](https://user-images.githubusercontent.com/59925371/114192676-0ed0dd00-994e-11eb-8ca8-12efdfd98561.png)

**Comment by utsav-akhaury** [comment]
Proper convergence is achieved for both [real](https://github.com/utsav-akhaury/ADMM/blob/main/FBS_real_UNet.ipynb) and [parametric](https://github.com/utsav-akhaury/ADMM/blob/main/FBS_param_UNet.ipynb) data when the U-Net is trained on a range of sigma values (instead of a constant value).
